### PR TITLE
[DataSource] Restrict to manage data source on the DSM UI

### PR DESCRIPTION
**Pull request by yubonluo**
### Description
Restrict to edit data source on the DSM UI.
manageableBy:
```
in workspace:
	anyone can not manage data source on the DSM
out of workspace:
	none: anyone can not manage data source on the DSM
	dashboard_admin: only OSD admin can manage source on the DSM
	all: anyone can manage data source on the DSM
in the saved object management page:
         all: anyone can delete/import data source
	dashboard_admin: only OSD admin can delete/import source
	none: anyone can not delete data source
```
### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6889

## Screenshot
manageableBy="all" or
manageableBy="dashboard_admin" and user is OSD admin
image
image
image

manageableBy="none" or
manageableBy="dashboard_admin" and user is not OSD admin or
user in a workspace
image
image
image

## Testing the changes
1. Clone the latest osd/security-dashboard plugin code and run yarn osd bootstrap
2. Modify config/opensearch_dashboards.yml
```yml
opensearchDashboards.dashboardAdmin.groups: ["dashboard_admin"]
opensearchDashboards.dashboardAdmin.users: ["username"]
workspace.enabled: true 
opensearch.username: "username"
opensearch.password: "password" 
opensearch.ssl.verificationMode: none
savedObjects.permission.enabled: true
data_source.enabled: true
data_source.manageableBy: "all|dashboard_admin|none"
```
3. Run yarn start --no-base-path
4. Operate data source.

## Changelog
- feat: [DataSource] Restrict to edit data source on the DSM UI.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data_source/server/plugin.ts [hunk 1]
```diff
@@ -81,6 +83,26 @@ export class DataSourcePlugin implements Plugin<DataSourcePluginSetup, DataSourc
       dataSourceSavedObjectsClientWrapper.wrapperFactory
     );
 
+    const { editMode } = config;
+    core.capabilities.registerProvider(() => ({
+      dataSource: {
+        canEdit: false,
+        canAssign: false,
```
Comment (ruanyl) [inline src/plugins/data_source/server/plugin.ts:90]: Do we need `canAssign`? seems it's not used

Comment (yubonluo) [reply src/plugins/data_source/server/plugin.ts:90]: The 'canAssign' is reserved for @raintygao.

File: config/opensearch_dashboards.yml [hunk 2]
```diff
@@ -321,6 +321,12 @@
 #   AWSSigV4:
 #     enabled: true
 
+# Optional setting that controls the permissions of data source to create, update and delete.
+# "read_only": The data source is readonly for all users including OSD admin.
+# "admin_only": The data source can only be edited by OSD admin.
+# "none": The data source can be edited by all users. Default to "none".
+# data_source.editMode: "none"
```
Comment (wanglam) [inline config/opensearch_dashboards.yml:328]: @ruanyl @SuZhou-Joe What do you think about renaming `editMode` to `operateMode`? The `editMode` is a little bit confused to me. Seems editMode only control the data source edit behavior. We still can create new data source.

Comment (SuZhou-Joe) [reply config/opensearch_dashboards.yml:328]: What about:
```
data_source.manageable_by: "all"
data_source.manageable_by: "dashboard_admin"
data_source.manageable_by: "none"
```

and default to `all` to keep backward compatible?

Comment (ruanyl) [reply config/opensearch_dashboards.yml:328]: @wanglam That's good point, agree with that.
I think what @SuZhou-Joe suggested sounds good to me.

File: src/plugins/data_source/server/plugin.ts (hunk 1 above)
Comment (ruanyl) [reply src/plugins/data_source/server/plugin.ts:90]: I don't think we should maintain this flag in data source plugin, it sounds better to be moved to workspace plugin.

**Review by ruanyl — DISMISSED** [review]

File: src/plugins/data_source/server/plugin.ts [hunk 3]
```diff
@@ -81,6 +83,25 @@ export class DataSourcePlugin implements Plugin<DataSourcePluginSetup, DataSourc
...
+      const { requestWorkspaceId, isDashboardAdmin } = getWorkspaceState(request);
+      // User can not manage data source in the workspace.
+      const canManage =
+        (manageableBy === ManageableBy.All && !requestWorkspaceId) ||
+        (manageableBy === ManageableBy.DashboardAdmin &&
+          isDashboardAdmin !== false &&
+          !requestWorkspaceId);
+
+      return { dataSource: { canManage } };
+    });
```
Comment (SuZhou-Joe) [inline src/plugins/data_source/server/plugin.ts:103]: Should we move this switcher to workspace?

File: src/plugins/data_source_management/public/components/edit_data_source/components/header/header.tsx [hunk 4]
```diff
@@ -175,11 +177,15 @@ export const Header = ({
       <EuiFlexItem grow={false}>
         <EuiFlexGroup alignItems="baseline" gutterSize="m" responsive={false}>
           {/* Test default button */}
-          <EuiFlexItem grow={false}>{renderDefaultIcon()}</EuiFlexItem>
+          {canManageDataSource ? (
+            <EuiFlexItem grow={false}>{renderDefaultIcon()}</EuiFlexItem>
+          ) : null}
```
Comment (SuZhou-Joe) [inline src/plugins/data_source_management/public/components/edit_data_source/components/header/header.tsx:182]: Should we restrict user from setting default data source?

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Pull request merged**